File: src/main/java/me/lokka30/levelledmobs/rules/strategies/YDistanceStrategy.java
/*
 * Copyright (c) 2020-2021  lokka30. Use of this source code is governed by the GNU AGPL v3.0 license that can be found in the LICENSE.md file.
 */

package me.lokka30.levelledmobs.rules.strategies;

import me.lokka30.levelledmobs.misc.LivingEntityWrapper;
import org.jetbrains.annotations.NotNull;

import java.util.concurrent.ThreadLocalRandom;

/**
 * Holds the configuration and logic for applying a levelling system that is
 * based upon the distance y level height
 *
 * @author stumper66
 * @since 3.0.0
 */
public class YDistanceStrategy implements LevellingStrategy, Cloneable {
    public Integer startingYLevel;
    public Integer endingYLevel;
    public Integer yPeriod;

    public void mergeRule(final LevellingStrategy levellingStrategy){
        if (levellingStrategy instanceof YDistanceStrategy)
            mergeYDistanceStrategy((YDistanceStrategy) levellingStrategy);
    }

    private void mergeYDistanceStrategy(final YDistanceStrategy yds){
        if (yds == null) return;

        if (yds.startingYLevel != null) this.startingYLevel = yds.startingYLevel;
        if (yds.endingYLevel != null) this.endingYLevel = yds.endingYLevel;
        if (yds.yPeriod != null) this.yPeriod = yds.yPeriod;
    }

    public String toString(){
        return String.format("ycoord: start: %s, end: %s, yPeriod: %s",
                startingYLevel == null ? 0 : startingYLevel,
                endingYLevel == null ? 0 : endingYLevel,
                yPeriod == null ? 0 : yPeriod
        );
    }

    public int generateLevel(@NotNull final LivingEntityWrapper lmEntity, final int minLevel, final int maxLevel) {

        final int mobYLocation = lmEntity.getLivingEntity().getLocation().getBlockY();
        final int yStart = this.startingYLevel == null ? 0 : this.startingYLevel;
        final int yEnd = this.endingYLevel == null ? 0 : this.endingYLevel;
        final double yPeriod = this.yPeriod == null ? 0.0 : this.yPeriod;
        int useLevel;
        final double diff = yEnd - yStart;

        if (yPeriod > 0) {
            final double percent = (mobYLocation - yStart) / diff;
            final double lvlPerPeriod = (maxLevel - minLevel) / (diff / yPeriod);
            useLevel = (int) Math.floor(minLevel + (lvlPerPeriod * (mobYLocation - yStart) / yPeriod));
        } else {
            final double useMobYLocation = mobYLocation - yStart;
            final double percent = useMobYLocation / diff;
            useLevel = minLevel + (int) Math.ceil((maxLevel - minLevel) * percent);
        }

        useLevel += getVariance(lmEntity, useLevel >= maxLevel);

        if (useLevel < minLevel)
            useLevel = minLevel;
        else if (useLevel > maxLevel)
            useLevel = maxLevel;

        return useLevel;
    }

    private int getVariance(@NotNull final LivingEntityWrapper lmEntity, final boolean isAtMaxLevel){
        final int variance = lmEntity.getMainInstance().rulesManager.getRule_MaxRandomVariance(lmEntity);
        if (variance == 0) return 0;

        final int change = ThreadLocalRandom.current().nextInt(0, variance + 1);

        // Start variation. First check if variation is positive or negative towards the original level amount.
        if (!isAtMaxLevel || ThreadLocalRandom.current().nextBoolean()) {
            // Positive. Add the variation to the final level
            return change;
        } else {
            // Negative. Subtract the variation from the final level
            return -change;
        }
    }

    public YDistanceStrategy cloneItem() {
        YDistanceStrategy copy = null;
        try {
            copy = (YDistanceStrategy) super.clone();
        } catch (final Exception e) {
            e.printStackTrace();
        }

        return copy;
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/rules/strategies/YDistanceStrategy.kt
package io.github.arcaneplugins.levelledmobs.rules.strategies

import java.util.concurrent.ThreadLocalRandom
import io.github.arcaneplugins.levelledmobs.LevelledMobs
import io.github.arcaneplugins.levelledmobs.wrappers.LivingEntityWrapper
import kotlin.math.floor

/**
 * Holds the configuration and logic for applying a levelling system that is based upon the distance
 * y level height
 *
 * @author stumper66
 * @since 3.0.0
 */
class YDistanceStrategy : LevellingStrategy, Cloneable {
    var startingYLevel: Int? = null
    var endingYLevel: Int? = null
    var yPeriod: Int? = null
    var increasePerLevel: Float? = null

    override val strategyType = StrategyType.Y_COORDINATE
    override var shouldMerge: Boolean = false

    override fun mergeRule(levellingStrategy: LevellingStrategy?) {
        if (levellingStrategy is YDistanceStrategy)
            mergeYDistanceStrategy(levellingStrategy as YDistanceStrategy?)
    }

    private fun mergeYDistanceStrategy(yds: YDistanceStrategy?) {
        if (yds == null) return

        if (yds.startingYLevel != null)
            this.startingYLevel = yds.startingYLevel

        if (yds.endingYLevel != null)
            this.endingYLevel = yds.endingYLevel

        if (yds.yPeriod != null)
            this.yPeriod = yds.yPeriod

        if (yds.increasePerLevel != null)
            this.increasePerLevel = yds.increasePerLevel
    }

    override fun toString(): String {
        return String.format(
            "y coord, start: %s, end: %s, yPeriod: %s, increasePerLvl: %s",
            if (startingYLevel == null) 0 else startingYLevel,
            if (endingYLevel == null) 0 else endingYLevel,
            if (yPeriod == null) 0 else yPeriod,
            if (increasePerLevel == null) 0 else increasePerLevel
        )
    }

    override fun generateNumber(
        lmEntity: LivingEntityWrapper,
        minLevel: Int,
        maxLevel: Int
    ): Float {
        var mobYLocation = lmEntity.livingEntity.location.blockY.toFloat()
        val yStart = if (this.startingYLevel == null) 0f else startingYLevel!!.toFloat()
        val yEnd = if (this.endingYLevel == null) 0f else endingYLevel!!.toFloat()
        val yPeriod = if (this.yPeriod == null) 0f else yPeriod!!.toFloat()
        val useLevel: Float
        val isDescending = (yStart > yEnd)

        val highest = if (isDescending) yStart else yEnd
        val lowest = if (isDescending) yEnd else yStart
        val diff = if (isDescending) yStart - yEnd else yEnd - yStart

        // make sure the mob location isn't past the end or start
        if (isDescending && yPeriod == 0f && increasePerLevel == null)
            mobYLocation = mobYLocation.coerceAtMost(highest)

        if (yPeriod == 0f && increasePerLevel == null)
            mobYLocation = mobYLocation.coerceAtLeast(lowest)

        val distanceBelow = if (isDescending)
            highest - mobYLocation
        else
            mobYLocation - lowest

        if (increasePerLevel != null){
            val firstStep = if (isDescending)
                yStart - mobYLocation
            else
                mobYLocation - yStart

            useLevel = floor(firstStep / increasePerLevel!!)
        }
        else if (yPeriod != 0f) {
            val lvlPerPeriod = (maxLevel - minLevel) / yPeriod
            val periodBelow = distanceBelow / yPeriod
            useLevel = minLevel + (lvlPerPeriod * periodBelow)
        } else {
            val useMobYLocation = distanceBelow.toDouble()
            val percent = (useMobYLocation / diff).toFloat()
            useLevel = minLevel + ((maxLevel - minLevel) * percent)
        }

        return useLevel
    }

    private fun getVariance(
        lmEntity: LivingEntityWrapper,
        isAtMaxLevel: Boolean
    ): Int {
        val variance = LevelledMobs.instance.rulesManager.getRuleMaxRandomVariance(
            lmEntity
        )
        if (variance == null || variance == 0)
            return 0

        val change = ThreadLocalRandom.current().nextInt(0, variance + 1)

        // Start variation. First check if variation is positive or negative towards the original level amount.
        return if (!isAtMaxLevel || ThreadLocalRandom.current().nextBoolean()) {
            // Positive. Add the variation to the final level
            change
        } else {
            // Negative. Subtract the variation from the final level
            -change
        }
    }

    override fun cloneItem(): YDistanceStrategy {
        var copy: YDistanceStrategy? = null
        try {
            copy = super.clone() as YDistanceStrategy
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return copy!!
    }
}